File: day_6/main.py
def get_first_start(marker_size: int) -> int:
    buffer = open("input.txt").read().strip()
    for offset in range((len(buffer) - marker_size) + 1):
        limit = offset + marker_size
        if len(set(buffer[offset:limit])) == marker_size:
            return limit
    return 0


def get_first_start_of_packet():
    packet_marker_size = 4
    return get_first_start(packet_marker_size)


print(get_first_start_of_packet())
assert get_first_start_of_packet() == 1723


def get_first_start_of_message():
    message_marker_size = 14
    return get_first_start(message_marker_size)


print(get_first_start_of_message())
assert get_first_start_of_message() == 3708

File: day_6/main.go
package main

import (
	"fmt"
	"os"
	"strings"
)

const (
	PacketMarkerSize  = 4
	MessageMarkerSize = 14
)

type Set map[rune]bool

func newSet(input string) Set {
	set := make(Set)

	for _, r := range input {
		set[r] = true
	}

	return set
}

func getFirstStart(buffer string, markerSize int) int {
	for offset := 0; offset < len(buffer)-markerSize; offset++ {
		limit := offset + markerSize
		set := newSet(buffer[offset:limit])
		if len(set) == markerSize {
			return limit
		}

	}
	return 0
}

func main() {
	data, _ := os.ReadFile("input.txt")
	buffer := strings.TrimRight(string(data), "\n")
	fmt.Println(getFirstStart(buffer, PacketMarkerSize))
	fmt.Println(getFirstStart(buffer, MessageMarkerSize))
}
